lib/storage/src/content_manager/consensus/consensus_wal.rs

```diff
@@ -74,13 +78,171 @@ impl ConsensusOpWal {
+        Ok(offset)
+    }
+
     pub fn append_entries(&mut self, entries: Vec<RaftEntry>) -> Result<(), StorageError> {
         for entry in entries {
-            log::debug!("Appending entry: {entry:?}");
+            let operation_opt = ConsensusOperations::try_from(&entry).ok();
+
+            let index = entry.index;
+            let current_index = self.0.last_index();
+            let index_offset = self.index_offset()?;
+
+            match index_offset {
+                Some(offset) => {
+                    // Assume we can't skip index numbers in WAL except for snapshot
+                    // Example: 2 <= 0 + 1 + 1
+                    debug_assert!(
+                        index <= current_index + offset + 1,
+                        "Expected no index skip: {} <= {} + {}",
+                        index,
+                        current_index,
+                        offset
+                    );
+
+                    if index < current_index + offset {
+                        // If there is a conflict, example:
+                        // Offset = 1
+                        // raft index = 10
+                        // wal index = 11
+                        // expected_wal_index = 10 - 1 = 9
+                        // 10 < 11 + 1
+                        if index < offset {
+                            return Err(StorageError::service_error(&format!(
+                                "Wal index conflict, raft index: {}, wal index: {}, offset: {}",
+                                index, current_index, offset
+                            )));
+                        }
+                        log::debug!(
+                            "Truncate conflicting WAL entries from index {}, raft: {}",
+                            index - offset,
+                            index
+                        );
+                        self.0.truncate(index - offset)?;
+                    } // else:
+                      // Offset = 1
+                      // raft index = 11
+                      // wal index = 10
+                      // expected_wal_index = 11 - 1 = 10
+                      // 11 < 10 + 1
+                }
+                None => {
+                    // There is no offset => there are no records in WAL
+                    // If there are no records, conflict is impossible
+                    // So we do nothing
+                }
+            }
+
+            if let Some(operation) = operation_opt {
+                let term = entry.term;
+                log::debug!(
+                    "Appending operation: term: {term}, index: {index} {operation:?}, entry: "
+                );
+            } else {
+                log::debug!("Appending entry: {entry:?}");
+            }
+
             let mut buf = vec![];
             entry.encode(&mut buf)?;
-            self.0.append(&buf)?;
+            let _wal_index = self.0.append(&buf)?;
```

Comment:
Let's do a `debug_assert` here?

Reply:
it's a non-trivial check, it depends on if there is a current offset

Reply:
Yeah, I understand. We can do something like that:
```rust
#[cfg(debug_assertions)]
if ... {
   debug_assert!(...)
} else {
   debug_assert!(...)
}

Reply:
:heavy_check_mark: